File: backend/app/services/rate_limit.py
from __future__ import annotations

import hashlib
import os
import time
from dataclasses import dataclass
from typing import Any

import httpx
from starlette.requests import Request


@dataclass(frozen=True)
class RateLimitWindow:
    name: str
    limit: int
    seconds: int


@dataclass(frozen=True)
class RateLimitResult:
    allowed: bool
    configured: bool
    limit: int
    remaining: int
    reset_seconds: int
    retry_after_seconds: int


RATE_LIMIT_GROUPS: dict[str, list[RateLimitWindow]] = {
    "chat": [
        RateLimitWindow("minute", 10, 60),
    ],
    "quant": [
        RateLimitWindow("minute", 60, 60),
        RateLimitWindow("day", 1000, 86400),
    ],
    "mf-detail": [
        RateLimitWindow("minute", 60, 60),
        RateLimitWindow("day", 1000, 86400),
    ],
    "category-funds": [
        RateLimitWindow("minute", 60, 60),
        RateLimitWindow("day", 1000, 86400),
    ],
    "fund-research": [
        RateLimitWindow("minute", 10, 60),
        RateLimitWindow("day", 200, 86400),
    ],
    "search": [
        RateLimitWindow("minute", 30, 60),
        RateLimitWindow("day", 500, 86400),
    ],
    "data-health": [
        RateLimitWindow("minute", 30, 60),
        RateLimitWindow("day", 500, 86400),
    ],
    "cron-sync-mf": [
        RateLimitWindow("hour", 2, 3600),
    ],
    "admin-mutation": [
        RateLimitWindow("minute", 20, 60),
    ],
}

_memory_store: dict[str, tuple[int, float]] = {}


def _enabled() -> bool:
    return os.getenv("RATE_LIMIT_ENABLED", "true").strip().lower() not in {"0", "false", "no", "off"}


def _upstash_config() -> tuple[str, str, bool]:
    url = os.getenv("UPSTASH_REDIS_REST_URL", "").strip().rstrip("/")
    token = os.getenv("UPSTASH_REDIS_REST_TOKEN", "").strip()
    return url, token, bool(url and token)


def _is_production() -> bool:
    env_values = [
        os.getenv("APP_ENV", ""),
        os.getenv("ENVIRONMENT", ""),
        os.getenv("NODE_ENV", ""),
        os.getenv("RENDER_ENV", ""),
    ]
    return any(value.strip().lower() == "production" for value in env_values) or bool(
        os.getenv("RENDER") or os.getenv("RENDER_SERVICE_ID")
    )


def _hash(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:24]


def client_identifier_from_request(request: Request, override: str | None = None) -> str:
    if override:
        return _hash(str(override))
        
    cf_ip = str(request.headers.get("cf-connecting-ip") or "").strip()
    real_ip = str(request.headers.get("x-real-ip") or "").strip()
    
    forwarded_ips = str(request.headers.get("x-forwarded-for") or "").split(",")
    forwarded = forwarded_ips[-1].strip() if forwarded_ips and forwarded_ips[0] else ""

    raw = (
        cf_ip
        or real_ip
        or forwarded
        or (request.client.host if request.client else "")
        or "unknown"
    )
    return _hash(raw)


def _window_key(group: str, identity: str, window: RateLimitWindow, now_seconds: int) -> str:
    bucket = now_seconds // window.seconds
    return f"rl:{group}:{identity}:{window.name}:{bucket}"


def _seconds_until_reset(window: RateLimitWindow, now_seconds: int) -> int:
    next_reset = ((now_seconds // window.seconds) + 1) * window.seconds
    return max(next_reset - now_seconds, 1)


async def _read_memory_window(group: str, identity: str, window: RateLimitWindow, now_seconds: int) -> tuple[int, int, RateLimitWindow]:
    key = _window_key(group, identity, window, now_seconds)
    reset_seconds = _seconds_until_reset(window, now_seconds)
    now_monotonic = time.monotonic()
    count, expires_at = _memory_store.get(key, (0, 0.0))
    if expires_at <= now_monotonic:
        count = 0
    count += 1
    _memory_store[key] = (count, now_monotonic + reset_seconds)
    return count, reset_seconds, window


async def _read_upstash_window(group: str, identity: str, window: RateLimitWindow, now_seconds: int) -> tuple[int, int, RateLimitWindow]:
    url, token, _configured = _upstash_config()
    key = _window_key(group, identity, window, now_seconds)
    reset_seconds = _seconds_until_reset(window, now_seconds)
    async with httpx.AsyncClient(timeout=3.0) as client:
        response = await client.post(
            f"{url}/pipeline",
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
            },
            json=[
                ["INCR", key],
                ["EXPIRE", key, reset_seconds + 5],
            ],
        )
        response.raise_for_status()
        payload: Any = response.json()
    count = int((payload or [{}])[0].get("result") or 0)
    return count, reset_seconds, window


async def check_rate_limit(
    group: str,
    identity: str,
    *,
    now_seconds: int | None = None,
) -> RateLimitResult:
    windows = RATE_LIMIT_GROUPS.get(group)
    if not _enabled() or not windows:
        return RateLimitResult(True, True, 0, 0, 0, 0)

    _url, _token, configured = _upstash_config()
    use_memory = not configured and not _is_production()
    if not configured and not use_memory:
        return RateLimitResult(False, False, 0, 0, 60, 60)

    current_seconds = now_seconds if now_seconds is not None else int(time.time())
    reads = [
        await (
            _read_memory_window(group, identity, window, current_seconds)
            if use_memory
            else _read_upstash_window(group, identity, window, current_seconds)
        )
        for window in windows
    ]
    most_limited = min(reads, key=lambda item: item[2].limit - item[0])
    blocked = [item for item in reads if item[0] > item[2].limit]
    allowed = not blocked
    retry_after = max((item[1] for item in blocked), default=0)

    count, reset_seconds, window = most_limited
    return RateLimitResult(
        allowed=allowed,
        configured=True,
        limit=window.limit,
        remaining=max(window.limit - count, 0),
        reset_seconds=reset_seconds,
        retry_after_seconds=retry_after,
    )


def rate_limit_headers(result: RateLimitResult) -> dict[str, str]:
    headers = {
        "X-RateLimit-Limit": str(result.limit),
        "X-RateLimit-Remaining": str(result.remaining),
        "X-RateLimit-Reset": str(result.reset_seconds),
    }
    if not result.allowed:
        headers["Retry-After"] = str(result.retry_after_seconds)
    return headers


def reset_rate_limit_memory_for_tests() -> None:
    _memory_store.clear()

File: frontend/lib/rateLimit.ts
import { createHash } from 'crypto';
import { UserRole, UserTier, effectiveRateLimitTier } from './billing/tiers';

type RateLimitWindow = {
  name: string;
  limit: number;
  seconds: number;
};

export type RateLimitGroup =
  | 'chat'
  | 'quant'
  | 'mf-detail'
  | 'category-funds'
  | 'search'
  | 'data-health'
  | 'feedback'
  | 'cron-sync-mf'
  | 'admin-mutation';

export type RateLimitResult = {
  allowed: boolean;
  configured: boolean;
  limit: number;
  remaining: number;
  resetSeconds: number;
  retryAfterSeconds: number;
};

export const RATE_LIMIT_GROUPS: Record<RateLimitGroup, RateLimitWindow[]> = {
  chat: [
    { name: 'minute', limit: 10, seconds: 60 },
  ],
  quant: [
    { name: 'minute', limit: 60, seconds: 60 },
    { name: 'day', limit: 1000, seconds: 86400 },
  ],
  'mf-detail': [
    { name: 'minute', limit: 60, seconds: 60 },
    { name: 'day', limit: 1000, seconds: 86400 },
  ],
  'category-funds': [
    { name: 'minute', limit: 60, seconds: 60 },
    { name: 'day', limit: 1000, seconds: 86400 },
  ],
  search: [
    { name: 'minute', limit: 30, seconds: 60 },
    { name: 'day', limit: 500, seconds: 86400 },
  ],
  'data-health': [
    { name: 'minute', limit: 30, seconds: 60 },
    { name: 'day', limit: 500, seconds: 86400 },
  ],
  feedback: [
    { name: 'minute', limit: 10, seconds: 60 },
    { name: 'day', limit: 100, seconds: 86400 },
  ],
  'cron-sync-mf': [
    { name: 'hour', limit: 2, seconds: 3600 },
  ],
  'admin-mutation': [
    { name: 'minute', limit: 20, seconds: 60 },
  ],
};

export const RATE_LIMIT_TIERS: Record<UserTier, Partial<Record<RateLimitGroup, RateLimitWindow[]>>> = {
  free: {
    chat: [
      { name: 'minute', limit: 5, seconds: 60 },
    ],
    quant: [
      { name: 'minute', limit: 20, seconds: 60 },
      { name: 'day', limit: 100, seconds: 86400 },
    ],
    'mf-detail': [
      { name: 'minute', limit: 20, seconds: 60 },
      { name: 'day', limit: 100, seconds: 86400 },
    ],
    'category-funds': [
      { name: 'minute', limit: 30, seconds: 60 },
      { name: 'day', limit: 300, seconds: 86400 },
    ],
    search: [
      { name: 'minute', limit: 10, seconds: 60 },
      { name: 'day', limit: 50, seconds: 86400 },
    ],
    'data-health': [
      { name: 'minute', limit: 15, seconds: 60 },
      { name: 'day', limit: 100, seconds: 86400 },
    ],
    feedback: [
      { name: 'minute', limit: 5, seconds: 60 },
      { name: 'day', limit: 30, seconds: 86400 },
    ],
  },
  pro: RATE_LIMIT_GROUPS,
  ultra: {
    chat: [
      { name: 'minute', limit: 30, seconds: 60 },
    ],
    quant: [
      { name: 'minute', limit: 180, seconds: 60 },
      { name: 'day', limit: 3000, seconds: 86400 },
    ],
    'mf-detail': [
      { name: 'minute', limit: 180, seconds: 60 },
      { name: 'day', limit: 3000, seconds: 86400 },
    ],
    'category-funds': [
      { name: 'minute', limit: 180, seconds: 60 },
      { name: 'day', limit: 3000, seconds: 86400 },
    ],
    search: [
      { name: 'minute', limit: 90, seconds: 60 },
      { name: 'day', limit: 1500, seconds: 86400 },
    ],
    'data-health': [
      { name: 'minute', limit: 90, seconds: 60 },
      { name: 'day', limit: 1500, seconds: 86400 },
    ],
    feedback: [
      { name: 'minute', limit: 20, seconds: 60 },
      { name: 'day', limit: 300, seconds: 86400 },
    ],
    'cron-sync-mf': RATE_LIMIT_GROUPS['cron-sync-mf'],
    'admin-mutation': RATE_LIMIT_GROUPS['admin-mutation'],
  },
};

type WindowUsage = {
  count: number;
  resetSeconds: number;
  window: RateLimitWindow;
};

const memoryStore = new Map<string, { count: number; expiresAtMs: number }>();
let warnedInMemoryFallback = false;

function isEnabled(): boolean {
  return String(process.env.RATE_LIMIT_ENABLED ?? 'true').trim().toLowerCase() !== 'false';
}

function upstashConfig() {
  const url = String(process.env.UPSTASH_REDIS_REST_URL || '').trim().replace(/\/$/, '');
  const token = String(process.env.UPSTASH_REDIS_REST_TOKEN || '').trim();
  return { url, token, configured: Boolean(url && token) };
}

function hashValue(value: string): string {
  return createHash('sha256').update(value).digest('hex').slice(0, 24);
}

export function getClientIp(request: Request): string {
  const forwarded = request.headers.get('x-forwarded-for')?.split(',')[0]?.trim();
  return (
    forwarded ||
    request.headers.get('x-real-ip')?.trim() ||
    request.headers.get('cf-connecting-ip')?.trim() ||
    'unknown'
  );
}

export function rateLimitIdentity(request: Request, override?: string): string {
  return hashValue(String(override || getClientIp(request)));
}

function windowKey(group: RateLimitGroup, identity: string, window: RateLimitWindow, nowMs: number): string {
  const bucket = Math.floor(nowMs / 1000 / window.seconds);
  return `rl:${group}:${identity}:${window.name}:${bucket}`;
}

function secondsUntilReset(window: RateLimitWindow, nowMs: number): number {
  const nowSeconds = Math.floor(nowMs / 1000);
  const nextReset = (Math.floor(nowSeconds / window.seconds) + 1) * window.seconds;
  return Math.max(nextReset - nowSeconds, 1);
}

async function readMemoryWindow(group: RateLimitGroup, identity: string, window: RateLimitWindow, nowMs: number): Promise<WindowUsage> {
  const key = windowKey(group, identity, window, nowMs);
  const resetSeconds = secondsUntilReset(window, nowMs);
  const existing = memoryStore.get(key);
  if (!existing || existing.expiresAtMs <= nowMs) {
    memoryStore.set(key, { count: 1, expiresAtMs: nowMs + resetSeconds * 1000 });
    return { count: 1, resetSeconds, window };
  }
  existing.count += 1;
  return { count: existing.count, resetSeconds, window };
}

async function readUpstashWindow(group: RateLimitGroup, identity: string, window: RateLimitWindow, nowMs: number): Promise<WindowUsage> {
  const { url, token } = upstashConfig();
  const key = windowKey(group, identity, window, nowMs);
  const resetSeconds = secondsUntilReset(window, nowMs);
  const response = await fetch(`${url}/pipeline`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify([
      ['INCR', key],
      ['EXPIRE', key, resetSeconds + 5],
    ]),
    cache: 'no-store',
  });
  if (!response.ok) {
    throw new Error(`upstash_rate_limit_failed:${response.status}`);
  }
  const payload = await response.json();
  const count = Number(payload?.[0]?.result || 0);
  return { count, resetSeconds, window };
}

export async function checkRateLimit(
  request: Request,
  group: RateLimitGroup,
  options: { identifier?: string; nowMs?: number; tier?: UserTier; role?: UserRole } = {},
): Promise<RateLimitResult> {
  const tier = effectiveRateLimitTier(options.tier, options.role);
  const windows = RATE_LIMIT_TIERS[tier][group] || RATE_LIMIT_GROUPS[group];
  if (!isEnabled() || !windows) {
    return { allowed: true, configured: true, limit: 0, remaining: 0, resetSeconds: 0, retryAfterSeconds: 0 };
  }

  const config = upstashConfig();
  if (!config.configured) {
    if (process.env.NODE_ENV === 'production') {
      return { allowed: false, configured: false, limit: 0, remaining: 0, resetSeconds: 60, retryAfterSeconds: 60 };
    }
    if (!warnedInMemoryFallback) {
      warnedInMemoryFallback = true;
      console.warn('Rate limit storage (Upstash Redis) is not configured; falling back to in-memory rate limiting.');
    }
  }
  const useMemory = !config.configured;

  const nowMs = options.nowMs ?? Date.now();
  const identity = rateLimitIdentity(request, options.identifier);
  const usages = await Promise.all(
    windows.map((window) => (
      useMemory
        ? readMemoryWindow(group, identity, window, nowMs)
        : readUpstashWindow(group, identity, window, nowMs)
    )),
  );

  const mostLimited = usages.reduce((selected, usage) => {
    const selectedRemaining = selected.window.limit - selected.count;
    const usageRemaining = usage.window.limit - usage.count;
    return usageRemaining < selectedRemaining ? usage : selected;
  }, usages[0]);
  const allowed = usages.every((usage) => usage.count <= usage.window.limit);
  const retryAfterSeconds = allowed
    ? 0
    : Math.max(...usages.filter((usage) => usage.count > usage.window.limit).map((usage) => usage.resetSeconds));

  return {
    allowed,
    configured: true,
    limit: mostLimited.window.limit,
    remaining: Math.max(mostLimited.window.limit - mostLimited.count, 0),
    resetSeconds: mostLimited.resetSeconds,
    retryAfterSeconds,
  };
}

export function rateLimitHeaders(result: RateLimitResult): Headers {
  const headers = new Headers();
  headers.set('X-RateLimit-Limit', String(result.limit));
  headers.set('X-RateLimit-Remaining', String(result.remaining));
  headers.set('X-RateLimit-Reset', String(result.resetSeconds));
  if (!result.allowed) {
    headers.set('Retry-After', String(result.retryAfterSeconds));
  }
  return headers;
}

export function rateLimitResponse(result: RateLimitResult): Response {
  const status = result.configured ? 429 : 503;
  const error = result.configured ? 'rate_limited' : 'rate_limit_unconfigured';
  return Response.json(
    { error, retry_after_seconds: result.retryAfterSeconds },
    { status, headers: rateLimitHeaders(result) },
  );
}

export async function enforceRateLimit(
  request: Request,
  group: RateLimitGroup,
  options: { identifier?: string; tier?: UserTier; role?: UserRole } = {},
): Promise<Response | null> {
  let result: RateLimitResult;
  try {
    result = await checkRateLimit(request, group, options);
  } catch {
    result = { allowed: false, configured: false, limit: 0, remaining: 0, resetSeconds: 60, retryAfterSeconds: 60 };
  }
  return result.allowed ? null : rateLimitResponse(result);
}

export function resetRateLimitMemoryForTests(): void {
  memoryStore.clear();
}
